app/models/concerns/foreman/sti.rb

```diff
@@ -12,8 +12,9 @@ module STI
     module ClassMethods
       # ensures that the correct STI object is created when :type is passed.
       def new_with_cast(*attributes, &block)
-        if (h = attributes.first).is_a?(Hash) && (type = h.with_indifferent_access.delete(:type)) && type.length > 0
-          if (klass = type.constantize) != self
+        default_type = self.respond_to?(:default_sti_class) && self.default_sti_class
```

Comment:
Do you mean add tests for this concern at all? I added tests for the behaviour in ```test/unit/subnet.rb```.